### ISSUE TITLE
when using react-hot-loader and typescript throws errors

### ISSUE THREAD

**Issue by veeramarni** [sysgears/spinjs#46]
Noticed when using `react-hot-loader` as well as `ts` in the stack, it throws following error. 

Stack tried
```
        stack: [
            "apollo",
            "ts",
            "react",
            "webpack",
            "react-hot-loader",
            "css"
        ],

```

Error
--
```
spin error TypeError: Cannot read property 'loader' of undefined
    at ReactHotLoaderPlugin.configure (/Users/user/Documents/common-stack/node_modules/spinjs/src/plugins/ReactHotLoaderPlugin.ts:24:38)
    at /Users/user/Documents/common-stack/node_modules/spinjs/src/createConfig.ts:88:62
    at Array.forEach (<anonymous>)
    at _loop_1 (/Users/user/Documents/common-stack/node_modules/spinjs/src/createConfig.ts:88:21)
    at Object.createConfig [as default] (/Users/user/Documents/common-stack/node_modules/spinjs/lib/createConfig.js:110:9)
    at Object.<anonymous> (/Users/user/Documents/common-stack/node_modules/spinjs/src/cli.ts:37:28)
    at Module._compile (module.js:643:30)
    at Object.Module._extensions..js (module.js:654:10)
    at Module.load (module.js:556:32)
    at tryModuleLoad (module.js:499:12)
    at Function.Module._load (module.js:491:3)
    at Module.require (module.js:587:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/user/Documents/common-stack/node_modules/spinjs/spin.js:4:1)
    at Module._compile (module.js:643:30)
    at Object.Module._extensions..js (module.js:654:10)
```

Spin config
---
```
spin debug SpinJS Config:
 { 'servers/backend-server[server]':
   { entry: './src/index.ts',
     stack:
      Stack {
        technologies: [ 'apollo', 'ts', 'css', 'webpack', 'server' ],
        platform: 'server' },
     tsLoaderOptions: { configFileName: './tsconfig.json' },
     defines:
      { __DEV__: true,
        __GRAPHQL_URL__: '"http://localhost:8080/graphql"',
        __SSR__: false,
        __PERSIST_GQL__: false,
        __FRONTEND_BUILD_DIR__: '"../frontend-server/dist/build"',
        __DLL_BUILD_DIR__: '"../frontend-server/dist/build/dll"',
        __SERVER__: true },
     enabled: true,
     name: 'server',
     require:
      { [Function: require]
        resolve: [Function],
        probe: [Function],
        cwd: '/Users/veeramarni/Documents/eclipse/workspace/phoneapps/development/projects/IDE/adminIde-stack/servers/backend-server' },
     plugins:
      [ WebpackPlugin {},
        WebAssetsPlugin {},
        CssProcessorPlugin {},
        ApolloPlugin {},
        TypeScriptPlugin {},
        BabelPlugin {},
        ReactPlugin {},
        ReactHotLoaderPlugin {},
        TCombPlugin {},
        FlowRuntimePLugin {},
        ReactNativePlugin {},
        ReactNativeWebPlugin {},
        StyledComponentsPlugin {},
        AngularPlugin {},
        VuePlugin {} ],
     roles: [ 'build', 'watch' ],
     cache: '../../.cache',
     ssr: false,
     backendBuildDir: 'dist',
     frontendBuildDir: 'dist',
     dllBuildDir: 'dist/.build/dll',
     webpackDll: false,
     reactHotLoader: false,
     persistGraphQL: false,
     backendUrl: 'http://{ip}:8080',
     frontendRefreshOnBackendChange: true,
     nodeDebugger: false,
     externals: [ [Function], [Function] ],
     devProxy: false,
     id: 'servers/backend-server[server]',
     buildDir: undefined,
     sourceMap: true },
  'servers/backend-server[test]':
   { stack:
      Stack {
        technologies: [ 'apollo', 'ts', 'css', 'webpack', 'server' ],
        platform: 'server' },
     roles: [ 'test' ],
     defines:
      { __DEV__: true,
        __GRAPHQL_URL__: '"http://localhost:8080/graphql"',
        __SSR__: false,
        __PERSIST_GQL__: false,
        __FRONTEND_BUILD_DIR__: '"../frontend-server/dist/build"',
        __DLL_BUILD_DIR__: '"../frontend-server/dist/build/dll"',
        __TEST__: true },
     name: 'test',
     require:
      { [Function: require]
        resolve: [Function],
        probe: [Function],
        cwd: '/Users/veeramarni/Documents/eclipse/workspace/phoneapps/development/projects/IDE/adminIde-stack/servers/backend-server' },
     plugins:
      [ WebpackPlugin {},
        WebAssetsPlugin {},
        CssProcessorPlugin {},
        ApolloPlugin {},
        TypeScriptPlugin {},
        BabelPlugin {},
        ReactPlugin {},
        ReactHotLoaderPlugin {},
        TCombPlugin {},
        FlowRuntimePLugin {},
        ReactNativePlugin {},
        ReactNativeWebPlugin {},
        StyledComponentsPlugin {},
        AngularPlugin {},
        VuePlugin {} ],
     cache: '../../.cache',
     ssr: false,
     backendBuildDir: 'dist',
     frontendBuildDir: 'dist',
     dllBuildDir: 'dist/.build/dll',
     webpackDll: false,
     reactHotLoader: false,
     persistGraphQL: false,
     backendUrl: 'http://{ip}:8080',
     frontendRefreshOnBackendChange: true,
     nodeDebugger: false,
     externals: [ [Function], [Function] ],
     devProxy: false,
     id: 'servers/backend-server[test]',
     buildDir: undefined,
     sourceMap: true },
  'servers/frontend-server[web]':
   { entry: './src/index.tsx',
     stack:
      Stack {
        technologies:
         [ 'apollo',
           'ts',
           'react',
           'webpack',
           'react-hot-loader',
           'css',
           'web' ],
        platform: 'web' },
     tsLoaderOptions: { configFileName: './tsconfig.json' },
     webpackDevPort: 3000,
     openBrowser: true,
     defines:
      { __DEV__: true,
        __GRAPHQL_URL__: '"http://localhost:8080/graphql"',
        __SSR__: false,
        __PERSIST_GQL__: '\'false\'',
        __FRONTEND_BUILD_DIR__: '\'dist\'',
        __DLL_BUILD_DIR__: '\'dist/.build/dll\'',
        __CLIENT__: true },
     htmlTemplate: '../../tools/html-plugin-template.ejs',
     waitOn: [ 'tcp:localhost:8080' ],
     enabled: true,
     name: 'web',
     require:
      { [Function: require]
        resolve: [Function],
        probe: [Function],
        cwd: '/Users/veeramarni/Documents/eclipse/workspace/phoneapps/development/projects/IDE/adminIde-stack/servers/frontend-server' },
     plugins:
      [ WebpackPlugin {},
        WebAssetsPlugin {},
        CssProcessorPlugin {},
        ApolloPlugin {},
        TypeScriptPlugin {},
        BabelPlugin {},
        ReactPlugin {},
        ReactHotLoaderPlugin {},
        TCombPlugin {},
        FlowRuntimePLugin {},
        ReactNativePlugin {},
        ReactNativeWebPlugin {},
        StyledComponentsPlugin {},
        AngularPlugin {},
        VuePlugin {} ],
     roles: [ 'build', 'watch' ],
     cache: '../../.cache',
     backendBuildDir: 'dist',
     frontendBuildDir: 'dist',
     dllBuildDir: 'dist/.build/dll',
     ssr: false,
     backendUrl: 'http://localhost:8080',
     webpackDll: true,
     reactHotLoader: true,
     persistGraphQL: false,
     frontendRefreshOnBackendChange: true,
     nodeDebugger: false,
     overridesConfig: './tools/webpackAppConfig.js',
     devProxy: false,
     id: 'servers/frontend-server[web]',
     buildDir: undefined,
     sourceMap: true },
  'servers/frontend-server[server]':
   { entry: './src/backend/app.ts',
     stack:
      Stack {
        technologies:
         [ 'apollo',
           'ts',
           'react',
           'webpack',
           'react-hot-loader',
           'css',
           'server' ],
        platform: 'server' },
     tsLoaderOptions: { configFileName: './tsconfig.json' },
     defines:
      { __DEV__: true,
        __GRAPHQL_URL__: '"http://localhost:8080/graphql"',
        __SSR__: false,
        __PERSIST_GQL__: '\'false\'',
        __FRONTEND_BUILD_DIR__: '\'dist\'',
        __DLL_BUILD_DIR__: '\'dist/.build/dll\'',
        __SERVER__: true },
     enabled: false,
     name: 'server',
     require:
      { [Function: require]
        resolve: [Function],
        probe: [Function],
        cwd: '/Users/veeramarni/Documents/eclipse/workspace/phoneapps/development/projects/IDE/adminIde-stack/servers/frontend-server' },
     plugins:
      [ WebpackPlugin {},
        WebAssetsPlugin {},
        CssProcessorPlugin {},
        ApolloPlugin {},
        TypeScriptPlugin {},
        BabelPlugin {},
        ReactPlugin {},
        ReactHotLoaderPlugin {},
        TCombPlugin {},
        FlowRuntimePLugin {},
        ReactNativePlugin {},
        ReactNativeWebPlugin {},
        StyledComponentsPlugin {},
        AngularPlugin {},
        VuePlugin {} ],
     roles: [ 'build', 'watch' ],
     cache: '../../.cache',
     backendBuildDir: 'dist',
     frontendBuildDir: 'dist',
     dllBuildDir: 'dist/.build/dll',
     ssr: false,
     backendUrl: 'http://localhost:8080',
     webpackDll: true,
     reactHotLoader: true,
     persistGraphQL: false,
     frontendRefreshOnBackendChange: true,
     nodeDebugger: false,
     overridesConfig: './tools/webpackAppConfig.js',
     devProxy: false,
     id: 'servers/frontend-server[server]',
     buildDir: undefined,
     sourceMap: true },
  'servers/frontend-server[test]':
   { stack:
      Stack {
        technologies:
         [ 'apollo',
           'ts',
           'react',
           'webpack',
           'react-hot-loader',
           'css',
           'server' ],
        platform: 'server' },
     roles: [ 'test' ],
     defines:
      { __DEV__: true,
        __GRAPHQL_URL__: '"http://localhost:8080/graphql"',
        __SSR__: false,
        __PERSIST_GQL__: '\'false\'',
        __FRONTEND_BUILD_DIR__: '\'dist\'',
        __DLL_BUILD_DIR__: '\'dist/.build/dll\'',
        __TEST__: true },
     name: 'test',
     require:
      { [Function: require]
        resolve: [Function],
        probe: [Function],
        cwd: '/Users/veeramarni/Documents/eclipse/workspace/phoneapps/development/projects/IDE/adminIde-stack/servers/frontend-server' },
     plugins:
      [ WebpackPlugin {},
        WebAssetsPlugin {},
        CssProcessorPlugin {},
        ApolloPlugin {},
        TypeScriptPlugin {},
        BabelPlugin {},
        ReactPlugin {},
        ReactHotLoaderPlugin {},
        TCombPlugin {},
        FlowRuntimePLugin {},
        ReactNativePlugin {},
        ReactNativeWebPlugin {},
        StyledComponentsPlugin {},
        AngularPlugin {},
        VuePlugin {} ],
     cache: '../../.cache',
     backendBuildDir: 'dist',
     frontendBuildDir: 'dist',
     dllBuildDir: 'dist/.build/dll',
     ssr: false,
     backendUrl: 'http://localhost:8080',
     webpackDll: true,
     reactHotLoader: true,
     persistGraphQL: false,
     frontendRefreshOnBackendChange: true,
     nodeDebugger: false,
     overridesConfig: './tools/webpackAppConfig.js',
     devProxy: false,
     id: 'servers/frontend-server[test]',
     buildDir: undefined,
     sourceMap: true } }
```

**Comment by vlasenko** [comment]
@veeramarni `react-hot-loader` has never worked well for us, if you or someone else want to fix it - PR are welcome, otherwise most probably it won't be fixed

**Comment by veeramarni** [comment]
Ok, may we need to use React Hot Loader v4 outside `spinjs` as it doesn't need [patch](https://github.com/gaearon/react-hot-loader#no-patch-required). 
I have tried it before but throws a harmless warning when using with ReactRouter. Not sure what exactly the warning is but I will try and if it is working then let you know.

**Comment by veeramarni** [comment]
@vlasenko 

I have some luck with the latest version of `React-Hot-Loader`. I was able to fix the other issue I was having. And I created a pull request. With `hot` api it is easier to implement it outside `spinjs`. But I haven't tested it thoroughly but it works for us so far and I tried with `Apollo Starter Kit` which worked as well and I only tested the web part by commenting anything related to `react-native`.

Created PR just to show where I made changes. 
https://github.com/sysgears/apollo-universal-starter-kit/pull/686

![image](https://user-images.githubusercontent.com/3382355/38222222-77aad0fe-36b1-11e8-8ec8-cb51d73c491a.png)
